### ISSUE TITLE
'Save Button' not working

### ISSUE THREAD

**Issue by Tasha53505** [Tasha53505/159.356-capstone-project-group#35]
I need to find out where the "save settings' are being sent to, because as of now - nothing is saving. i.e languages etc.

**Comment by Tasha53505** [comment]
saving is done through .prefs files insidee C:\ProgramData\Squeezebox\prefs  folder. I will review the code first but basically the /en passess the information to the LMS backend and the backend when acknowledged that the setting is clean, it writes it to e C:\ProgramData\Squeezebox\prefs\<specificname>.prefs

what you can do (and what I do) is not let the LMS backend handle checking it, I go directly to the prefs folder and write it myself through a script

open the <>.prefs folders read how the settings are saved and then you can figure out if you can copy how /en gives the settings to the backend OR bypass the backend and write a separate settings manager program that handles writing to the prefs folder

**Comment by Tasha53505** [comment]
With LMS, there are rules about writing plugins, I think any perl scripts might need to be inside the Plugins folder and clearly defined perl scripts before they can activate, I would ask the dev comunity if you can inject custom perl code into LMS or if you must adhere to the plugins folder


[When running my script locally ](https://github.com/Tasha53505/159.356-capstone-project-group/commit/dd41c0603f96c36d79d4e466437500f3d7b1a0f0) upon restart, the server DOES update with language, but not

**Comment by Tasha53505** [comment]
I have submitted an LMS forum.

**Comment by Tasha53505** [comment]
https://forums.slimdevices.com/forum/user-forums/logitech-media-server/1725784-trouble-getting-lms-backend-to-recognize-perl-script

LMS Forum

**Comment by Tasha53505** [comment]
Updated ReadMe with plugin api documentation

**Comment by Tasha53505** [comment]
Now,
When running my script locally:
```
package Plugins::LanguageSettings;
use strict;
use File::Slurp qw(read_file write_file);
use Slim::Utils::Prefs;

# Create a preferences object for the plugin
my $prefs = preferences('plugin.languagesettings');

# Define the path to the prefs file
my $prefs_file_path = 'C:/ProgramData/Squeezebox/prefs/server.prefs';

# Subroutine to get the display name for the plugin in the menu
sub getDisplayName {
    return 'PLUGIN_LANGUAGE_SETTINGS';
}

# Return strings for localization
sub strings {
    return {
        'PLUGIN_LANGUAGE_SETTINGS' => 'Language Settings',
        'EN'                       => 'English',
        'DE'                       => 'German',
        'FR'                       => 'French',
    };
}

# Subroutine to update the language setting in the prefs file
sub update_language {
    my ($new_language) = @_;

     # Hardcode the language to French for testing
    my $new_language = 'FR';

    # Read the current content of the prefs file
    my $file_data = read_file($prefs_file_path);

    # Modify the line that starts with 'language:'
    $file_data =~ s/^language: \w+/language: $new_language/m;

    # Write the updated content back to the prefs file
    write_file($prefs_file_path, $file_data);

    # Update the plugin preferences with the new language
    $prefs->set('language', $new_language);
}

# Lines function: this can be used to display messages on the LMS UI
sub lines {
    my ($client) = @_;
    
    # Return a hash as per the new display API
    return {
        'line1' => $client->symbols('PLUGIN_LANGUAGE_SETTINGS'),
        'line2' => 'Language updated successfully',
    };
}

# Set the mode and display the plugin output
sub setMode {
    my ($class, $client) = @_;
    $client->lines(\&lines);
    $client->update();
}

# Add the plugin to the menu
sub addMenu {
    return "PLUGINS";
}

# Define the functions for handling remote button events
sub initPlugin {
    my ($class) = @_;
    
    %functions = (
        'left'  => sub { my $client = shift; Slim::Buttons::Common::popModeRight($client); },
        'up'    => sub { my $client = shift; $client->bumpUp(); },
        'down'  => sub { my $client = shift; $client->bumpDown(); },
    );

    Slim::Buttons::Common::addSaver($class);  # Register screensaver if applicable
}

sub getFunctions {
    return \%functions;
}

# Execute the update_language function based on input from the UI
sub handleRequest {
    my ($request) = @_;
    my $client    = $request->client;
    my $language  = $request->getParam('language');

    if (defined $language) {
        update_language($language);
        setMode(__PACKAGE__, $client);
    }
}

1;

```

With a hardcoded value of FR --> 
`    my $new_language = 'FR';`

## ERROR:
Can't locate Slim/Utils/Prefs.pm in @INC (you may need to install the Slim::Utils::Prefs module) (@INC entries checked: C:/Strawberry/perl/site/lib C:/Strawberry/perl/vendor/lib C:/Strawberry/perl/lib) at LanguageSettings.pm line 4.
BEGIN failed--compilation aborted at LanguageSettings.pm line 4.

**Comment by Tasha53505** [comment]
**From the forum post** --> https://forums.slimdevices.com/forum/user-forums/logitech-media-server/1725784-trouble-getting-lms-backend-to-recognize-perl-script

If you want to rearrange how the settings are grouped, then you'll indeed have to write quite a bit of code: either you indeed create a plugin with your own page handlers (which will require Perl), or you do things client side as much as possible, using JSON/RPC to set the prefs. Both approaches have pros and cons.

I'd recommend you start simple to get a better understanding of how LMS works. Start by adding a new skin in the main code. This way you inherit all the page handlers, default styling etc. Then add some styling tweaks and see how far you get there. Look at the page handlers for the settings in Slim::Web::Settings to learn how those pages are handled. Read [https://lyrion.org/reference/cli/usi...cli/#jsonrpcjs](https://lyrion.org/reference/cli/using-the-cli/#jsonrpcjs) to see how you can set prefs from the frontend.

And then, when you need help, share your GitHub repo, so we can easily see what you did.

**Comment by Tasha53505** [comment]
https://lyrion.org/reference/cli/using-the-cli/#jsonrpcjs --> updated link

**Comment by Tasha53505** [comment]
Look at commit 

[git commit -m "Added jsonRPC with some basic code. Hardcoding French language to test SAVE button. also moved saveButton OUTSIDE of accordion, as it wasnt triggering my console log. This seems to be a separate problem"
](https://github.com/Tasha53505/159.356-capstone-project-group/commit/7b2b9f4bc0bc93c6b2c1d025c3247473a9081241)

**Comment by Tasha53505** [comment]
Hardcoded values upon SAVE do now work. - still NOT using the accordion save button. Now to work on the NON-hardcoded language values and using the accordion save button
-----------
I have gotten the "save settings function" - hardcoded language value, BUT SAVING and UPDATING functionality working. (A restart to the server IS required upon saving) I spent the last couple hours reading up on the documentation (which has been added to our documentation) - created a JSON  / RPC file and a function call for the Save Settings button. 

I'll now work on the non-hardcoded language value.

COMMIT: https://github.com/Tasha53505/159.356-capstone-project-group/commit/7f997c8e8000afe885f33b29f0fc425315f0aabb

**Comment by Tasha53505** [comment]
The save button ONLY works, if it's outside of the settingsContainer class. I am looking into this

**Comment by Tasha53505** [comment]
# The issue
The entire issue of the button not being clicked was because of elements being adding to the DOM after a page load. It was such an issue that not even Console Logs were triggering on my test buttons - and only in certain places. It took me ages, but I finally figured it out.

# The Fix

It meant that any elements added to the DOM after a page load, my event listeners attached directly to those elements didn't work because they were not present when the listeners were initially set up. (but visually the buttons APPEAR to click)

Not ONLY did I need
`
document.addEventListener("DOMContentLoaded", function() {
`

but **also** 
`document.body` (the overall parent elelment) which made it so that ANYthing that was clicked, even future elements added to the DOM would work because body is the most high up parent. In more technical jargon, I used event delegation.


```
document.addEventListener("DOMContentLoaded", function() {
    
    // Event delegation to handle dynamic content
    document.body.addEventListener('click', function(e) {
        if (e.target && e.target.id === 'testButton') {
            console.log('testButton clicked');
        } 
        
        if (e.target && e.target.id === 'saveSettings') {
            e.preventDefault(); // Prevent default behavior
            
            const selectedLanguage = document.getElementById("languageSelect").value;
            console.log("Save Settings BUTTON CLICKED");

            // Send a JSON-RPC request to update the language in the server.prefs file
            updateLanguageSetting(selectedLanguage);    
        }
    });
});
```
This enabled my test button to work (to actually log to the console) and most importantly, for my saveSettings Button to trigger)

## TLDR 
i.e  The fix was event delegation, by using the highest parent (great great great grandparent), the body element to listen for events - it could recognise events added to the DOM

## commit: https://github.com/Tasha53505/159.356-capstone-project-group/commit/62bc60748cd58cda4daf3af93c7cce2d67e5f750


### Still to do:
Still need to implement Non-hardcoded values, but that won't be an issue. Will work on that now before end of today

**Comment by Tasha53505** [comment]
Language now chnages upon restart and will correctly update with the language chosen. MUST restart server for it to apply
commit: https://github.com/Tasha53505/159.356-capstone-project-group/commit/aae3b39b6ebb90e354235485e7298dcd68d0c89c

**Comment by Tasha53505** [comment]
All progress and fixes above. Can close now
